### ISSUE TITLE
libcalico-go should improve error handle when use list+watch resource mode to reduce load on apiserver

### ISSUE THREAD

**Issue by gaopeiliang** [projectcalico/calico#8277]
we use an old version before https://github.com/projectcalico/libcalico-go/pull/1484 that list resource with consistent read;  it will read from etcd directly; 

we use felixconfig change some param,  too many list pod make our apiserver over load and oom ;

we study the master source code,  even though list + watch read from cache (rv=0) after  https://github.com/projectcalico/libcalico-go/pull/1484;  but I think some items should improve also;

1.  allowBookMark Should add like https://github.com/projectcalico/libcalico-go/pull/1484 TODO
2.  felixconfig change param should restart Felix Should Doc annotation; It is too rish in BigCluster;  When update config restart Should add random wait time reduce cluster load;
3.  golang use http2 default, so list + watch retry will not LB to another apiserver instance;  (over load with more retry)
     (1)  libcalico-go should use backoff when handle error like k8s goclient informer, eg: 429, service unavaliable;
     (2)  client config Burst add to 100,  client will send more request when some error;
     (3)  RVTooLarge error should handle like RV too small like https://github.com/projectcalico/calico/pull/6045
4.  data encode use application/vnd.kubernetes.protobuf will faster 10x  like kube-scheduler and kube-controller
5.  I am not sure is it necessary to load all Pod in Felix,  program NetworkPolicy ?    
   
@caseydavenport @fasaxc

**Comment by fasaxc** [comment]
Thanks for the tips, yes, it'd be good to get some of those done (they've been on the backlog for a while but it's hard to prioritise them vs more serious bugs).

You should use Typha if your APi server is getting overloaded.  That reduces the load because only Typha has a watch; felix just connects to Typha instead.

**Comment by gaopeiliang** [comment]
> Thanks for the tips, yes, it'd be good to get some of those done (they've been on the backlog for a while but it's hard to prioritise them vs more serious bugs).
> 
> You should use Typha if your APi server is getting overloaded. That reduces the load because only Typha has a watch; felix just connects to Typha instead.

explain about apiserver overload ,   it is one of cluster instance  overload because retry will always happen on One TCP Connection due to HTTP2; 

The process maybe :
1.   one instance maybe overload oom because too many request;
2.   all broken client reconnect to instance2,  overload ,  retry , oom;   
3.   the apiserver broken one by one ,  until all cluster ~~~~

If request LB to every instance ,  The server cap is enough;   But  http2 retry on One Instance will broken this One first;
 
 I think it is same as Typha ,  will overload one by one .... 
 
 
 you are right that maybe have more serious Todo;  I think calico is good enough,  But this Process just maybe happen on BigCluster ,  it is more important than  all small cluster;  expect more attention;  (smile)

**Comment by gaopeiliang** [comment]
> How many nodes do you have? Small clusters run 3 x Typha. If you have 100 nodes you'll have 3 connections to API server, not 100. 30x reduction in load.

you are right;   But I think  it just overload  from apiserver  to Typha Server Instance ....

**Comment by fasaxc** [comment]
Sorry, I don't understand; Typha won't get overloaded, it's designed to handle hundreds of Felix connections.  If your API server can't handle 3 typha connections, perhaps your API server is too small?

**Comment by fasaxc** [comment]
Please upgrade Calico and Kubernetes to the latest versions and then start using Typha.  If that doesn't solve your problem we'll be in a much better position to investigate.

**Comment by gaopeiliang** [comment]
we will try it when have an opportunity;  

I think main is  HTTP2  multi stream on One TCP Connection ,  L4 lb about connection has no affect about stream requests; 

kube-apiserver use --goaway-chance  random close http2 connection to make connection more blance ,  even though same connection != same http request ,  but have an change to reblance connection; 

It is same as other Servers use HTTP2  list + watch mode ....

**Comment by gaopeiliang** [comment]
> Watch bookmarks is the main one that would really help, I think. It's just very tricky to plumb through libcalico-go without breaking anything else. There are several layers with very subtle error handling that need to be dealt with.

Get data with encode with protobuf from apiserver  

Encode from json to protobuf breaking nothing when get data from apiserver also;

just:

```
config, _ := clientcmd.BuildConfigFromFlags("", *kubeconfig)
config.ContentType = "application/vnd.kubernetes.protobuf"
```

**Comment by fasaxc** [comment]
I'd love to give that a go if it has such a big impact; I worry about subtle breakages/incompatibilities with all our custom CRDs and so on but maybe it just works and gives a big win.

Why don't you put up a PR and we can run it through our CI tests.

**Comment by fasaxc** [comment]
Looks like protobuf only supported for some resources: https://kubernetes.io/docs/reference/using-api/api-concepts/#alternate-representations-of-resources.  That makes me more confident that it's a good idea: only a few resources to check and no CRD changes to worry about.

**Comment by gaopeiliang** [comment]
> I'd love to give that a go if it has such a big impact; I worry about subtle breakages/incompatibilities with all our custom CRDs and so on but maybe it just works and gives a big win.
> 
> Why don't you put up a PR and we can run it through our CI tests.

maybe I can have a try ......

**Comment by gaopeiliang** [comment]
I think Typha is an better design (send batch msg to other component with TCP Conn) after study source code  to reduce apiserver load ....

**Comment by fasaxc** [comment]
Typha and Felix share the libcalico-go code, any improvement to one also improves the other (but simply switching to Typha should reduce load a lot just by virtual of having 100x fewer connections to your API servers).

**Comment by gaopeiliang** [comment]
> Typha and Felix share the libcalico-go code, any improvement to one also improves the other (but simply switching to Typha should reduce load a lot just by virtual of having 100x fewer connections to your API servers).

yes

**Comment by fasaxc** [comment]
@gaopeiliang are you sure we don't use protobuf?  I asked on Kubernetes slack and they said client-go uses it by default

**Comment by fasaxc** [comment]
Breadcrumb numbers aren't revisions, they're not in sync between typhas.  Clients start from scratch with the new typha connection.  Recent versions of typha support compressed snapshots and they share them between clients.  Much faster than the API server.

**Comment by gaopeiliang** [comment]
> Breadcrumb numbers aren't revisions, they're not in sync between typhas. Clients start from scratch with the new typha connection. Recent versions of typha support compressed snapshots and they share them between clients. Much faster than the API server.

en ,,  I think the obj in breadcrumb, like pod update obj  is versioned;    if  client scratch from diff typha instance that unsync ,  client maybe receive repeate breadcrumb, and repeate update obj ?  like pod obj version from 10 -> 7 -> 8 -> 9 -> 10

**Comment by fasaxc** [comment]
Calico is designed to be memoryless, it would apply state 10, then state 7, then state 8, then 9, then 10 again.  Not ideal but this is also very rare in practice because Typha is generally a couple of orders of magnitude faster than its clients.
